hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+              container.getContainerId());
+          queuedOpportunisticContainers.put(
+              container.getContainerId(), container);
+        } else {
+          isQueued = false;
+          LOG.info("Opportunistic container [{}] will not be queued at the NM" +
+              "since max queue length [{}] has been reached",
+              container.getContainerId(), maxOppQueueLength);
+          container.sendKillEvent(
+              ContainerExitStatus.KILLED_BY_CONTAINER_SCHEDULER,
+              "Opportunistic container queue is full.");
+        }
+      }
+      if (isQueued) {
+        try {
+          this.context.getNMStateStore().storeContainerQueued(
+              container.getContainerId());
+        } catch (IOException e) {
+          LOG.warn("Could not store container state into store..", e);
+        }
+      }
+    }
+  }
+
+  private void killOpportunisticContainers(Container container) {
+    List<Container> extraOpportContainersToKill =
+        pickOpportunisticContainersToKill(container.getContainerId());
+    // Kill the opportunistic containers that were chosen.
+    for (Container contToKill : extraOpportContainersToKill) {
+      contToKill.sendKillEvent(
+          ContainerExitStatus.KILLED_BY_CONTAINER_SCHEDULER,
+          "Container Killed to make room for Guaranteed Container.");
+      oppContainersMarkedForKill.put(contToKill.getContainerId(), contToKill);
+      LOG.info(
+          "Opportunistic container {} will be killed in order to start the "
+              + "execution of guaranteed container {}.",
+          contToKill.getContainerId(), container.getContainerId());
+    }
+  }
+
+  private void startAllocatedContainer(Container container) {
+    LOG.info("Starting container [" + container.getContainerId()+ "]");
+    scheduledToRunContainers.put(container.getContainerId(), container);
+    this.utilizationManager.addContainerResources(container);
+    if (container.getContainerTokenIdentifier().getExecutionType() ==
+        ExecutionType.OPPORTUNISTIC) {
+      this.opportunisticContainersStatus.setOpportMemoryUsed(
+          this.opportunisticContainersStatus.getOpportMemoryUsed()
+              + container.getResource().getMemorySize());
+      this.opportunisticContainersStatus.setOpportCoresUsed(
+          this.opportunisticContainersStatus.getOpportCoresUsed()
+              + container.getResource().getVirtualCores());
+      this.opportunisticContainersStatus.setRunningOpportContainers(
+          this.opportunisticContainersStatus.getRunningOpportContainers()
+              + 1);
+    }
+    container.sendLaunchEvent();
+  }
+
+  private List<Container> pickOpportunisticContainersToKill(
+      ContainerId containerToStartId) {
+    // The additional opportunistic containers that need to be killed for the
+    // given container to start.
+    List<Container> extraOpportContainersToKill = new ArrayList<>();
```

Comment:
In this case, the 'extra' is needed. It signifies all the container above the already marked 'oppContainersToKill' that need to be killed.